nomos-services/data-availability/network/src/backends/libp2p/validator.rs

```diff
@@ -0,0 +1,217 @@
+use libp2p::PeerId;
+use log::error;
+use nomos_da_network_core::protocols::sampling;
+use nomos_da_network_core::protocols::sampling::behaviour::SamplingError;
+use nomos_da_network_core::swarm::validator::{ValidatorEventsStream, ValidatorSwarm};
+use nomos_da_network_core::SubnetworkId;
+use overwatch_rs::overwatch::handle::OverwatchHandle;
+use overwatch_rs::services::state::NoState;
+use std::fmt::Debug;
+use std::marker::PhantomData;
+use std::pin::Pin;
+use subnetworks_assignations::MembershipHandler;
+use tokio::sync::broadcast;
+use tokio::sync::mpsc::error::SendError;
+use tokio::sync::mpsc::UnboundedSender;
+use tokio::task::JoinHandle;
+use tokio_stream::wrappers::BroadcastStream;
+
+type BlobId = [u8; 32];
+type ColumnIdx = u32;
+
+const BROADCAST_CHANNEL_SIZE: usize = 128;
+
+/// Message that the backend replies to
+#[derive(Debug)]
+pub enum DaNetworkMessage {
+    /// Quickstart a network sapling
+    RequestSample {
+        subnetwork_id: ColumnIdx,
+        blob_id: BlobId,
+    },
+}
+
+/// Events types to subscribe to
+/// * Sampling: Incoming sampling events [success/fail]
+/// * Incoming blobs to be verified
+#[derive(Debug)]
+pub enum DaNetworkEventKind {
+    Sampling,
+    Verifying,
+}
+
+/// Sampling events coming from da network
+#[derive(Debug, Clone)]
+pub enum SamplingEvent {
+    /// A success sampling
+    SamplingSuccess { blob_id: BlobId, blob: Box<DaBlob> },
+    /// A failed sampling error
+    SamplingError { error: SamplingError },
+}
+
+/// DA network incoming events
+#[derive(Debug)]
+pub enum DaNetworkEvent {
+    Sampling(SamplingEvent),
+    Verifying(Box<DaBlob>),
+}
+
+/// DA network backend for validators
+/// Internally uses a libp2p swarm composed of the [`ValidatorBehaviour`]
+/// It forwards network messages to the corresponding subscription channels/streams
+pub struct DaNetworkValidatorBackend<Membership> {
+    // TODO: this join handles should be cancelable tasks. We should add an stop method for
+    // the `NetworkBackend` trait so if the service is stopped the backend can gracefully handle open
+    // sub-tasks as well.
+    #[allow(dead_code)]
+    task: JoinHandle<()>,
+    #[allow(dead_code)]
+    replies_task: JoinHandle<()>,
+    sampling_request_channel: UnboundedSender<(SubnetworkId, BlobId)>,
+    sampling_broadcast_receiver: broadcast::Receiver<SamplingEvent>,
+    verifying_broadcast_receiver: broadcast::Receiver<DaBlob>,
+    _membership: PhantomData<Membership>,
+}
+
+#[derive(Clone, Debug)]
+pub struct DaNetworkValidatorBackendSettings<Membership> {
+    /// Identification key
+    key: Keypair,
+    /// Membership of DA network PoV set
+    membership: Membership,
+}
+
+impl<Membership> DaNetworkValidatorBackend<Membership> {
+    /// Send the sampling request to the underlying sampling behaviour
+    async fn handle_sample_request(&self, subnetwork_id: SubnetworkId, blob_id: BlobId) {
+        if let Err(SendError((subnetwork_id, blob_id))) =
```

Comment:
This is just handling the error case, is it not desirable to also handle the success case - maybe just with a debug log for now, so we can better track things until we are stable?

Reply:
The success means it arrived to the other side. It is an error because it should not happen and we cannot recover from it. It means that the service in the other side consuming this is down or erred. I could add a `trace!` log for success but I think it would clog the log system with useless information in this case.

Reply:
I am just imagining the situation where we are debugging sampling, and want to be sure that a message was sent (and then on the other side received, but debugging across process boundaries is often messy).

If we have any means already to make sure that we can reconstruct this information, then I agree we don't need any more logs.

Reply:
If we do not see this error is guaranteed it was sent. Then we would have to check on the other side if it arrived, and log if we have a receive error though.